File: conventions/review/ts/test/conformance.test.ts
// The TS review conformance replay: it runs the REAL TS review verb against the TS
// Recorder and asserts the captured operations equal the language-neutral vectors
// under protocol/conformance/vectors/review — the SAME JSON files the Go suite
// replays (FORMAT.md, ADR-0041). Passing the IDENTICAL setReview.json +
// setReviewApprove.json vectors the Go suite passes is the op-transcript
// co-equality proof (TASK-239 AC#2/AC#9).
//
// The comparison reproduces FORMAT.md exactly: order-sensitive ops; per-op op /
// subject / name / expectedRev equality; each payload compared as canonical JSON
// via the SDK's `canonical`.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { canonical, type JSONValue } from "@sextant/sdk";
import { setReview, type SetReviewInput } from "../src/index.js";
import { Recorder, type Op } from "./recorder.js";
import { reviewVectorsDir } from "./repoRoot.js";

// OpTranscriptVector is the on-disk vector shape (protocol/conformance/vector.go's
// OpTranscriptVector / FORMAT.md). input is opaque to the format; the verb decodes it.
interface OpTranscriptVector {
  epoch: number;
  convention: string;
  verb: string;
  description?: string;
  input: JSONValue;
  operations: Op[];
}

function loadReviewVectors(): Array<{ path: string; vector: OpTranscriptVector }> {
  const dir = reviewVectorsDir();
  const files = readdirSync(dir).filter((f) => f.endsWith(".json"));
  if (files.length === 0) {
    throw new Error(`no review vectors found under ${dir}`);
  }
  return files
    .sort()
    .map((f) => {
      const path = join(dir, f);
      return { path, vector: JSON.parse(readFileSync(path, "utf8")) as OpTranscriptVector };
    });
}

// runVerb dispatches (convention, verb) to the registered TS verb, run against the
// recorder. Only setReview exists; an unknown verb fails loudly (a vector no verb
// answers is a gap, not a pass — mirroring the Go runner).
async function runVerb(rec: Recorder, v: OpTranscriptVector): Promise<void> {
  if (v.convention !== "review") {
    throw new Error(`vector names convention ${JSON.stringify(v.convention)}, not review`);
  }
  switch (v.verb) {
    case "setReview": {
      const input = v.input as unknown as SetReviewInput;
      await setReview(rec, input);
      return;
    }
    default:
      throw new Error(`vector names verb ${JSON.stringify(v.verb)}, which is not registered`);
  }
}

// seedReview seeds the recorder with the prior bus state a setReview vector reads
// before it writes — the exact peer of the Go suite's seedReview. For an approve it
// also seeds the goal the artifact's proof relation backs, so the closed-loop
// transcript is reproduced. The seed does NOT appear in the transcript.
function seedReview(rec: Recorder, v: OpTranscriptVector): void {
  if (v.verb !== "setReview") return;
  const input = v.input as unknown as SetReviewInput;
  if (input.state === "approved") {
    rec.seedArtifact(
      input.name,
      { $type: "doc", title: "PR #42", relates: [{ goal: "g1", crit: "c1", kind: "proof" }], review: { state: "review" } },
      2,
    );
    rec.seedArtifact(
      "goal.g1",
      { $type: "goal", northstar: "ship it", criteria: [{ id: "c1", text: "merged", status: "in-progress" }] },
      5,
    );
    return;
  }
  rec.seedArtifact(input.name, { $type: "doc", title: "the brief", body: "x", review: { state: "review" } }, 3);
}

// canonicalPayload canonicalizes an op's payload under the FORMAT.md rule. An
// absent payload canonicalizes to "null" (so two ops that both omit a payload
// compare equal) — mirroring Go's Canonicalize(nil) == "null".
function canonicalPayload(p: JSONValue | undefined): string {
  return canonical(p === undefined ? null : p);
}

// assertOpsEqual compares the recorded transcript to the vector's, order-sensitive,
// each payload under the canonical-JSON rule — the TS peer of the Go runner's
// assertOpsEqual.
function assertOpsEqual(path: string, want: Op[], got: Op[]): void {
  assert.equal(
    got.length,
    want.length,
    `${path}: operation count\n want: ${JSON.stringify(want.map((o) => o.op))}\n got:  ${JSON.stringify(got.map((o) => o.op))}`,
  );
  for (let i = 0; i < want.length; i++) {
    const w = want[i]!;
    const g = got[i]!;
    assert.equal(g.op, w.op, `${path}: op[${i}].op`);
    assert.equal(g.subject ?? "", w.subject ?? "", `${path}: op[${i}] (${w.op}).subject`);
    assert.equal(g.name ?? "", w.name ?? "", `${path}: op[${i}] (${w.op}).name`);
    assert.equal(g.expectedRev ?? null, w.expectedRev ?? null, `${path}: op[${i}] (${w.op}).expectedRev`);
    const cw = canonicalPayload(w.payload);
    const cg = canonicalPayload(g.payload);
    assert.equal(cg, cw, `${path}: op[${i}] (${w.op}).payload mismatch\n want: ${cw}\n got:  ${cg}`);
  }
}

const vectors = loadReviewVectors();

test("the review conformance vectors are discovered", () => {
  assert.ok(vectors.length >= 1, "expected at least one review vector");
  assert.ok(
    vectors.some((v) => v.path.endsWith("setReviewApprove.json")),
    "expected the setReviewApprove.json vector the Go suite passes",
  );
});

for (const { path, vector } of vectors) {
  const rel = path.split("/").slice(-2).join("/");
  test(`review vector ${rel} replays to the identical operations`, async () => {
    assert.equal(vector.epoch, 1, "the shipped review vector is pinned to epoch 1");
    const rec = new Recorder();
    seedReview(rec, vector);
    await runVerb(rec, vector);
    assertOpsEqual(path, vector.operations, rec.operations());
  });
}

File: conventions/review/go/conformance_test.go
package review_test

import (
	"context"
	"encoding/json"
	"fmt"
	"path/filepath"
	"testing"

	review "github.com/love-lena/sextant/conventions/review/go"
	pconf "github.com/love-lena/sextant/protocol/conformance"
	conf "github.com/love-lena/sextant/sdk/conformance"
)

// The review convention's conformance suite. It plugs the REAL review verb into
// the TASK-183 seam: register it in a conformance.Registry, then ReplayVectors over
// the language-neutral vectors under protocol/conformance/vectors/review — the SAME
// JSON files the TS conv-review suite replays (FORMAT.md, ADR-0041), so the two are
// co-equal (TASK-239 AC#2/AC#9). Two vectors pin the two transcripts: a plain
// verdict (get -> update) and an approve that runs the closed loop (get -> update ->
// goal.get -> goal.update -> publish goal.update).

// vectorsDir is the protocol-owned vector root, relative to this test
// (conventions/review/go -> repo root -> protocol/conformance/vectors).
func vectorsDir() string {
	return filepath.Join("..", "..", "..", "protocol", "conformance", "vectors")
}

// fixedNow is the verdict timestamp the recorded verb stamps, so the merged review
// block and any closed-loop goal.update are byte-stable across runs. The live
// SetReview takes the real time via SetReviewInput.Now; the recorded verb pins it.
const fixedNow = "2026-06-19T00:00:00Z"

// setReviewVerb adapts review.SetReview to a conformance.Verb: decode the vector's
// input and run the verb against the recorder (conf.Ops's method set is a superset
// of review.Ops, so the recorder is the verb's Ops unchanged).
func setReviewVerb(ctx context.Context, ops conf.Ops, input json.RawMessage) error {
	var in review.SetReviewInput
	if err := json.Unmarshal(input, &in); err != nil {
		return fmt.Errorf("review conformance: decode setReview input: %w", err)
	}
	_, err := review.SetReview(ctx, ops, in)
	return err
}

func reviewRegistry() *conf.Registry {
	reg := conf.NewRegistry()
	reg.Register("review", "setReview", setReviewVerb)
	return reg
}

// seedReview seeds the recorder with the prior bus state a setReview vector reads
// before it writes. SetReview is a read-then-write verb (get the artifact, merge
// the review block, update); the seed mirrors the bus state it would find live and
// does not appear in the transcript. For an approve it also seeds the goal the
// artifact's proof relation backs, so the closed-loop transcript is recorded.
func seedReview(rec *conf.Recorder, v pconf.OpTranscriptVector) {
	if v.Verb != "setReview" {
		return
	}
	var in review.SetReviewInput
	_ = json.Unmarshal(v.Input, &in)
	if in.State == "approved" {
		// The approved artifact declares a proof relation backing goal g1 / crit c1,
		// at revision 2 (the value the artifact.update CAS's against).
		rec.SeedArtifact(in.Name, json.RawMessage(
			`{"$type":"doc","title":"PR #42","relates":[{"goal":"g1","crit":"c1","kind":"proof"}],"review":{"state":"review"}}`,
		), 2)
		// The goal it backs (c1 not yet met), at revision 5.
		rec.SeedArtifact("goal.g1", json.RawMessage(
			`{"$type":"goal","northstar":"ship it","criteria":[{"id":"c1","text":"merged","status":"in-progress"}]}`,
		), 5)
		return
	}
	// A plain verdict: a producer-marked doc at revision 3 (the CAS target).
	rec.SeedArtifact(in.Name, json.RawMessage(`{"$type":"doc","title":"the brief","body":"x","review":{"state":"review"}}`), 3)
}

// TestReviewConformance replays the review vectors against the real verb. With
// -update it (re)records the on-disk sample vectors, so a deliberate verb change is
// a one-command re-record plus a reviewed diff.
func TestReviewConformance(t *testing.T) {
	dir := vectorsDir()
	reg := reviewRegistry()

	if conf.Updating() {
		recordReviewVectors(t, dir)
	}
	conf.ReplayVectors(t, dir, reg, conf.WithSeed(seedReview))
}

// recordReviewVectors writes the sample review vectors from the registered verb
// (the -update path). Each vector pins one verb call's transcript; the seed
// supplies the prior bus state the read-then-write verb needs.
func recordReviewVectors(t *testing.T, dir string) {
	t.Helper()
	cases := []struct {
		file        string
		description string
		input       review.SetReviewInput
	}{
		{
			file: "setReview.json",
			description: "review: set a producer-marked doc to the 'changes' verdict — " +
				"artifact.get then artifact.update (CAS the record with the review block " +
				"merged in, every other field preserved). No closed loop (not an approve).",
			input: review.SetReviewInput{Name: "the-brief", State: "changes", By: "01OPERATOR", Now: fixedNow},
		},
		{
			file: "setReviewApprove.json",
			description: "review: approve a proof-bearing doc — artifact.get + artifact.update " +
				"persist the verdict, then the approve->met closed loop runs the goals " +
				"convention's single write path on the backed criterion: goal.get, " +
				"goal.update (CAS c1 to met), message.publish a goal.update on msg.topic.goals.",
			input: review.SetReviewInput{Name: "the-proof", State: "approved", By: "01OPERATOR", Now: fixedNow},
		},
	}
	for _, c := range cases {
		input, err := json.Marshal(c.input)
		if err != nil {
			t.Fatalf("marshal %s input: %v", c.file, err)
		}
		v, err := conf.RecordVector(1, "review", "setReview", c.description, input, setReviewVerb,
			func(rec *conf.Recorder) {
				seedReview(rec, pconf.OpTranscriptVector{Verb: "setReview", Input: input})
			})
		if err != nil {
			t.Fatalf("record %s: %v", c.file, err)
		}
		path := filepath.Join(dir, "review", c.file)
		if err := conf.WriteVector(path, v); err != nil {
			t.Fatalf("write %s: %v", c.file, err)
		}
		t.Logf("re-recorded %s", path)
	}
}
